apps/api/src/lib/flat-parse.ts

```diff
@@ -1,118 +1,52 @@
+ * - Multiple invalid fields (without request part): "Invalid request: name (expected string, received number); age (expected number, received string)"
+ * - Pure schema validation error: "Invalid request: passwords do not match"
  *
  * @template TSchema - The Zod schema type
  * @param schema - The Zod schema to validate against
  * @param input - The input data to validate
- * @returns A FlatParseResult with either success data or a concatenated error message
- *
- * @example
- * ```typescript
- * // Basic usage
- * const userSchema = z.object({
- *   name: z.string().min(1, "Name is required"),
- *   email: z.string().email("Invalid email"),
- *   age: z.number().min(18, "Must be 18 or older")
- * });
- *
- * const result = flatParse(userSchema, { name: "", email: "invalid", age: 16 });
- * if (!result.success) {
- *   console.log(result.error);
- *   // Output: "Name is required, Invalid email, Must be 18 or older"
- * }
- * ```
- *
- * @example
- * ```typescript
- * // In an Express route handler
- * app.post('/users', (req, res) => {
- *   const result = flatParse(UserSchema, req.body);
- *   if (!result.success) {
- *     return res.status(400).json({ error: result.error });
- *   }
- *
- *   // result.data is fully typed and validated
- *   const user = result.data;
- *   // ... process user
- * });
- * ```
- *
- * @example
- * ```typescript
- * // Comparison with Zod's safeParse
- * // Zod's approach:
- * const zodResult = userSchema.safeParse(input);
- * if (!zodResult.success) {
- *   // zodResult.error.issues is an array of detailed error objects
- *   const errors = zodResult.error.issues.map(issue =>
- *     `${issue.path.join('.')}: ${issue.message}`
- *   );
- *   console.log(errors);
- * }
- *
- * // This utility's approach:
- * const flatResult = flatParse(userSchema, input);
- * if (!flatResult.success) {
- *   // flatResult.error is a single string with all messages
- *   console.log(flatResult.error);
- * }
- * ```
+ * @param requestPart - Optional string to specify which part of the request is being validated (for error messages)
+ * @returns The validated data, fully typed as z.TypeOf<TSchema>
+ * @throws {ApiError} If validation fails, throws with a concatenated error message
  */
-function flatParse<TSchema extends z.ZodType>(
+export function flatParse<TSchema extends z.ZodType>(
   schema: TSchema,
   input: unknown,
-): FlatParseResult<z.TypeOf<TSchema>> {
-  const { success, error, data } = z.safeParse(schema, input);
-  if (success) {
-    return { success: true, data: data };
+  requestPart?: string,
+): z.infer<TSchema> {
+  const result = schema.safeParse(input);
+  if (result.success) {
+    return result.data;
   } else {
-    const message = error?.issues.map((i) => i.message).join(", ");
-    return { success: false, error: message };
+    const message = result.error.issues
+      .map((issue) => {
+        // Clean up the error message by removing redundant "Invalid input: " prefix, include the
```

Comment:
we'd get messages like `Invalid request: Invalid input: expected string, received undefined, Passwords do not match`. 

now, it's be a bit easier to read: `Invalid request: id (expected string, received undefined); password (passwords do not match)`